volume/mounts/mounts.go

```diff
@@ -75,14 +76,36 @@ type MountPoint struct {
 	// Specifically needed for containers which are running and calls to `docker cp`
 	// because both these actions require mounting the volumes.
 	active int
+
+	// SafePaths created by Setup that should be cleaned up before unmounting
+	// the volume.
+	safePaths []*safepath.SafePath
 }
 
-// Cleanup frees resources used by the mountpoint
+// Cleanup frees resources used by the mountpoint and cleans up all the paths
+// returned by Setup that hasn't been cleaned up by the caller.
 func (m *MountPoint) Cleanup() error {
 	if m.Volume == nil || m.ID == "" {
 		return nil
 	}
 
+	for _, p := range m.safePaths {
+		if !p.IsValid() {
+			continue
+		}
+
+		base, sub := p.SourcePath()
+		log.G(context.TODO()).WithFields(logrus.Fields{
+			"path":          p.Path(),
+			"sourceBase":    base,
+			"sourceSubpath": sub,
+		}).Warn("cleaning up SafePath that hasn't been cleaned up by the caller")
+
+		if err := p.Close(); err != nil {
+			return errors.Wrapf(err, "error cleaning up safepath: %v", p)
```

Comment:
This error condition is intriguing.
I'm not sure what to do here.
What do we expect the user to do in this situation? Seems like this would block the container from moving to a shutdown state.

Reply:
Hmm, maybe we can ignore it at this point and only log a warning. If the cleanup failure makes it impossible to unmount the base volume, the `m.Volume.Unmount` will fail anyway. And if that still succeeds, then perhaps the failure wasn't critical.

Also, if this has been reached, this means that the caller of `Setup` didn't call the cleanup on his own, which probably means that something went really wrong in some other place.